src/Utilities/FlowAnalysis/Analysis/PropertySetAnalysis/OperationMethodKey.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+
+namespace Analyzer.Utilities.FlowAnalysis.Analysis.PropertySetAnalysis
+{
+    /// <summary>
+    /// Operation and Method pair.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Used as a key to identify invocation operations, which may have different
+    /// underlying invoked methods, cuz of delegates or something.
+    /// </para>
+    /// <para>
+    /// Operations and Symbols don't seem to override GetHashCode() / Equals(), so
+    /// different instances representing the same code will be considered not equal.
```

Comment:
I am confused here. How are we getting different operations for same code, given we use the same semantic model, which is guaranteed to return same operations for same syntax node.

Reply:
It confused me too.  :-)

So...I had a bug in the DoNotUseInsecureDeserializerWithoutBinderBase.cs, with using a ConcurrentBag instead of HashSet for rootOperationsNeedingAnalysis.  That resulted in me calling .GetEnclosingControlFlowGraph() multiple times on same IOperation instance.  And it appears that creating the ControlFlowGraph clones IOperations?  If I look at a different CFGs created from the same IOperation instance, the CFGs' Blocks[x].Operations[y] have different GetHashCode() values.

Also the PointsToAnalysisResults and CopyToAnalysisResults had different hash codes.

Reply:
Cool, that helped me identifier a bug in our code. We cache the `{ RootOperation => CFG }` mapping in the analyzer driver, so if you invoked [http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,1454](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,1454) on the analysis context, it will return you the same CFG if asked multiple times. However, this API was added very recently (may be it is also not in the MS.CA package we reference currently in the repo). As a workaround we have this code: [https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Utilities/Extensions/IOperationExtensions.cs#L444](https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Utilities/Extensions/IOperationExtensions.cs#L444), which force creates a new CFG every time. So basically, your original approach was fine if all pieces were in place.

---
In reply to: [224192958](https://github.com/dotnet/roslyn-analyzers/pull/1838#discussion_r224192958) [](ancestors = 224192958)

Reply:
In the future, should we expect the CFG's IOperations to be the same instances as non-DFA analysis sees?

Reply:
No, this will never happen as the operations get lowered and no longer belong to the original operation tree. Original operations as part of a single operation tree rooted at the operation block's root operation, while the new operations within a basic block all have null parent, so even the operation tree structure changes significantly.

---
In reply to: [224302593](https://github.com/dotnet/roslyn-analyzers/pull/1838#discussion_r224302593) [](ancestors = 224302593)